rust/nasl-builtin-network/src/socket.rs

```diff
@@ -0,0 +1,731 @@
+pub struct NaslSockets {
+    handles: RwLock<Handles>,
+    interval: Option<RwLock<Interval>>,
+}
+
+impl NaslSockets {
+    fn open_udp(addr: &str, port: u16) -> Result<NaslSocket, FunctionErrorKind> {
+        let socket = UdpSocket::bind("0.0.0.0:0")?;
+        socket.connect(format!("{addr}:{port}"))?;
+        socket.set_read_timeout(Some(Duration::from_secs(1)))?;
+        Ok(NaslSocket::Udp(UDPConnection {
+            socket,
+            buffer: vec![],
+        }))
+    }
+
+    fn open_tcp(
+        ip: &str,
+        port: u16,
+        bufsz: Option<i64>,
+        timeout: Option<Duration>,
+        tls_config: Option<&TLSConfig>,
+    ) -> Result<NaslSocket, FunctionErrorKind> {
+        // Resolve Address and Port to SocketAddr
+        let sock = format!("{ip}:{port}").to_socket_addrs()?.next().ok_or(
+            FunctionErrorKind::WrongArgument(format!(
+                "the given address and port do not correspond to a valid address: {ip}:{port}"
+            )),
+        )?;
+        // Create Vec depending of buffer size
+        let buffer = if let Some(bufsz) = bufsz {
+            if bufsz > 0 {
+                Some(Vec::with_capacity(bufsz as usize))
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+
+        // Create socket for connection
+        let socket = if let Some(timeout) = timeout {
+            TcpStream::connect_timeout(&sock, timeout)?
+        } else {
+            TcpStream::connect(sock)?
+        };
+
+        // Create TLS Connection if requested
+        let tls_connection = match tls_config {
+            Some(config) => Some(
+                rustls::ClientConnection::new(
+                    Arc::new(config.config.clone()),
+                    config.server.clone(),
+                )
+                .map_err(|e| {
+                    FunctionErrorKind::Diagnostic(
+                        format!("Unable to establish TLS connection: {e}"),
+                        None,
+                    )
+                })?,
+            ),
+            None => None,
+        };
+
+        Ok(NaslSocket::Tcp(Box::new(TCPConnection {
+            socket,
+            _tls_connection: tls_connection,
+            _buffer: buffer,
+        })))
+    }
+
+    fn add(&self, socket: NaslSocket) -> usize {
+        let mut handles = self
+            .handles
+            .write()
+            .expect("Unable to access socket handles");
+        let ret;
+        if let Some(free) = handles.closed_fd.pop() {
+            handles.handles.insert(free, socket);
+            ret = free;
+        } else {
+            ret = handles.handles.len();
+            handles.handles.push(socket);
+        }
+        ret
+    }
+
+    /// Close a given file descriptor taken as an unnamed argument.
+    fn close(&self, r: &Register, _: &Context) -> Result<NaslValue, FunctionErrorKind> {
+        let args = r.positional();
+        let socket = match args.get(1) {
```

Comment:
must be `args.get(0)` otherwise:
When I execute:
```
# let socket = UdpSocket::bind("127.0.0.1:34254")?;
data = "foo bar";
soc = open_sock_udp(34254);
n = send(socket:soc, data: data);
display(n);
close(soc);
```

with `scannerctl execute --target 127.0.0.1 udp.nasl`

I get:
```
7
2024-07-08T12:28:55.913279Z  WARN scannerctl: script error, udp.nasl, line: 146, col: 151: close(soc);: close: expected 1 arguments but got 1
```

The close error is misleading and it happens regardless if there is udp server running or not.